src/abstractions/serialization/KiotaSerializer.Serialization.cs

```diff
@@ -25,11 +25,12 @@ public static partial class KiotaSerializer
     /// <typeparam name="T">Type of the object to serialize</typeparam>
     /// <param name="contentType">Content type to serialize the object to </param>
     /// <param name="value">The object to serialize.</param>
+    /// <param name="useBackingStore">Backing store keeps track of changes, setting this to false will give you the full item.</param>
     /// <returns>The serialized representation as a stream.</returns>
-    public static Stream SerializeAsStream<T>(string contentType, T value) where T : IParsable
+    public static Stream SerializeAsStream<T>(string contentType, T value, bool useBackingStore = true) where T : IParsable
     {
         using var writer = GetSerializationWriter(contentType, value);
-        writer.WriteObjectValue(string.Empty, value);
+        writer.WriteObjectValue(useBackingStore ? string.Empty : null, value);
```

Comment:
I'm not sure I follow how switching between string empty and null impacts the backing store behaviour here? Those seem like two different concerns.

Reply:
I tried this code, mind the `null` instead of `string.empty`. and it gives me back the entire item.

```csharp

    public static Stream? SerializeAsJsonStream<T>(this T? input) where T : IParsable
    {
        if (input == null) return null;
        var writer = new Microsoft.Kiota.Serialization.Json.JsonSerializationWriter(new Microsoft.Kiota.Serialization.Json.KiotaJsonSerializationContext());
        writer.WriteObjectValue(null, input);
        return writer.GetSerializedContent();
    }
```

Reply:
the null key is not what's changing things here. It's the fact that the writer does not have the proxies registered for the backing store. (you're creating a new writer) so it reverts to the default behaviour (serialize everything)
That's actually a better solution than what I was suggesting (flip the backing store flags all the way across the tree) but we need to make sure we implement it in a way that does not hardcode things with a specific writer.
One way to do that would be to assign null to the events receivers of the writer when the parameter is true.

Reply:
Oops, I was to quick.

This abstraction library has no clue about the actual serializer (it's in the abstraction library). Do you mean:

```csharp
        using var writer = GetSerializationWriter(contentType, value);
        if (!serializeOnlyChangedValues) {
            writer.OnAfterObjectSerialization = null;
            writer.OnBeforeObjectSerialization = null;
            writer.OnStartObjectSerialization = null;
        }
        writer.WriteObjectValue(string.Empty, value);
```

Reply:
I think this may be better. Resetting the hooks may cause unintentional side effects later on if the model is being re-used and one expects them to still  be there or added them in a custom way. 

This enables us to target situation where the backing store is not returning all the values.....
```cs
        if(!serializeOnlyChangedValues && value is IBackedModel backedModel)
        {
            backedModel.BackingStore.InitializationCompleted = false;// reset the initialization completed flag to ensure all properties are serialized
        }
```

Reply:
Great idea @andrueastman, have any idea how I would do that on a collection?

Reply:
right, but if we choose this approach (what I initially had suggested), we'll need to do it for the whole tree.
we'll also need to reset the value back after the serialization.

maybe mixing both solutions would be the better approach here if a hook is registered, register an additional one that flips things up.
This way we don't impact the original behaviour (which might or might not be backing store related) and fix the issue WITHOUT having to go through the whole tree ourselves.
Thoughts?

Reply:
If we want [this](https://github.com/microsoft/kiota-dotnet/pull/311#discussion_r1698549272) the extension methods have to go into the *Kiota.Serialization.Json* library, because that is where it's known.

The solution @andrueastman is suggesting is applied on the code above. No sure if the collection part is working though.

Reply:
> This way we don't impact the original behavior (which might or might not be backing store related) and fix the issue WITHOUT having to go through the whole tree ourselves.

I agree with @baywet here. If we want to reset the flag/return the store to original state, we should probably reset the `ReturnOnlyChangedValues` flag and NOT the `InitializationCompleted` (as this will clear the store state).

Another idea we could probably do is add a public boolean property to the [BackingStoreSerializationWriterProxyFactory](https://github.com/microsoft/kiota-dotnet/blob/main/src/abstractions/store/BackingStoreSerializationWriterProxyFactory.cs#L12) that defaults to true. Then the hooks would use it to do their current actions. Otherwise allow all properties to be serialized by doing the opposite action.

If that works, we would simply need to update the code whenever [GetSerializationWriter](https://github.com/microsoft/kiota-dotnet/blob/bc0404d400e977e41c0dc277a845d6bf977ef44b/src/abstractions/serialization/KiotaSerializer.Serialization.cs#L115) is called to do something like this.

```cs
        if(!serializeOnlyChangedValues && writer is BackingStoreSerializationWriterProxyFactory)
            BackingStoreSerializationWriterProxyFactory.ReturnOnlyChangedValues = false; //enable writing all props

        writer.WriteObjectValue(string.Empty, value);

        if(!serializeOnlyChangedValues && writer is BackingStoreSerializationWriterProxyFactory)
            BackingStoreSerializationWriterProxyFactory.ReturnOnlyChangedValues = true; // reset to original state
```

Reply:
@andrueastman You're not able to access properties inside an action you create in the base constructor.

But I've slightly modified the [SerializationWriterFactoryRegistration](https://github.com/microsoft/kiota-dotnet/pull/311/files#diff-f305f58e4f7d9e3de38c6908dba355140eabdd5970ff388c40677291adbf64d6) to enable this change

Reply:
> @andrueastman You're not able to access properties inside an action you create in the base constructor.

I see that now. The suggested approach look great.